### ISSUE TITLE
Large (number of files) `save` operations are extremely slow and resource intensive

### ISSUE THREAD

**Issue by mih** [datalad/datalad#4630]
I often create datasets with many files (100k is pretty normal). Those operations produce an enormous number of files. For roughly 80k tracked files I get 200k+ files/dirs in `.git/annex/objects` (for well known reasons), but also 200k+ files in `.git/objects/`. For reasons outlined in https://github.com/datalad/datalad/issues/3869 these operations are rather slow. I wonder if some sort of intermediate compaction can help here, but I have no concrete ideas.

**Comment by yarikoptic** [comment]
annex: see https://github.com/datalad/datalad/issues/32#issuecomment-71927742 and references there-in (I had not used any "in production" yet) for lightening up `.git/annex/objects` a bit.   For what could potentially be done in the future see/join discussion on elderly
- https://git-annex.branchable.com/todo/shorter_keys_through_better_encoding/ - removal of `KEY/` directory
- https://git-annex.branchable.com/todo/external_backends/ - custom external backends as a more general solution . We might still be doing some suppression of `gc` runs, which were initially introduced to not disturb GitPython's in memory git db - may be those contribute as well.

git: run `git gc --prune=now` to prune (I believe left from some `git-annex` operations) and repack `.git/objects/`

**Comment by mih** [comment]
The tuning is largely impossible to use, because it influences all clones -- we discussed that before. However, I was talking about compaction that happens while `save` is still running. After `save` I am already left with 9 files in `.git/objects` and 180MB (so all good), but while `save` is running, there are 400k+ files and lots of GB in that folder (this dataset had 490k files, and ultimately 1.2M inodes in `.git/annex/objects`).

**Comment by yarikoptic** [comment]
> The tuning is largely impossible to use, because it influences all clones -- we discussed that before.

Could you please remind me what is the problem them? AFAIK as long as you don't want to merge separate repos/datasets which were born with different tuning parameters - everything should be fine.

> I was talking about compaction that happens while save is still running

oh, missed that aspect, sorry.  I guess it is worth looking what `git` processes are ran by save, with what settings (e.g. disabling gc or not etc), and monitor which ones trigger explosion of unpacked objects?
For such large in the number of files repos it might be worth also check out split index mode: https://git-scm.com/docs/git-update-index  . It should commit faster but not sure if it would provide any positive effect on number of files created/left behind.

**Comment by mih** [comment]
I am staring at another attempt to save 132k files (few kilobytes each). The progressbar suggests that this will take between a half and a full day. Transferring 3x the size and number of this dataset onto the drives of the system I am running DataLad on (across the internet) took 20min via rsync.

**Comment by bpoldrack** [comment]
Collecting some first profiling pieces for further investigation:
All fresh datasets only differing in number of files. Files contain nothing but a counter. Running on laptop.

`datalad save -m "Saving 10k files"  22.40s user 8.38s system 135% cpu 22.800 total`
`datalad save -m "Saving 100k files"  1373.15s user 165.64s system 215% cpu 11:53.00 total`

Saving 10k files:
![save_10k](https://user-images.githubusercontent.com/10498301/86512367-df5ed400-be01-11ea-8dac-80ddafd018f8.png)

Saving 10k files:
![save_100k](https://user-images.githubusercontent.com/10498301/86512404-64e28400-be02-11ea-8e9d-61d4da4677c1.png)

**Comment by yarikoptic** [comment]
Please place those images within

```html
<details>
<summary>Figure caption</summary> 

The content

</details>
```
to reduce clutter etc.

Above profiles IIRC % of CPU time used by the process which were consumed... but datalad quite often stays at very low CPU %, while time is spent in waiting/interactions.
Idea: it might be informative to do statistic profiling (at the processes level) down from the datalad process to see what aspects take how long through the course of execution by reviewing [flame graphs](http://www.brendangregg.com/flamegraphs.html) ... may be using [perf](https://www.percona.com/blog/2019/11/20/profiling-software-using-perf-and-flame-graphs/)

**Comment by bpoldrack** [comment]
Absolutely agree that this needs a closer look than those graphs. However, they can give a quick and easy hint whether we do something obviously redundant etc. - so, just a first check whether things appear to be sane.

Will look into `perf`, but for now did what I planned for anyway: `line_profile` + git/annex level timing as to be seen below. I think it's quite telling already:

1. The most problematic performance issue is happening at git/annex level. This is because we call `git (annex) add` explicitly on all the files, that `save` figures are to be added.

Compare those invocations using explicit file list vs. just `.` on a dataset with 50k files (`1.txt 2.txt ... 50000.txt`):
<details>
<summary>Plain `git-add`:</summary>

This varies of course a bit, but the factor remains in all trials a ran.

```
❱ time ls [0-9]* | xargs git add                                                 
 ls --color=auto [0-9]*  0.14s user 0.12s system 2% cpu 12.237 total              
xargs git add  51.85s user 0.36s system 173% cpu 30.176 total
```
vs.
```
❱ time git add .
git add .  0.50s user 1.17s system 96% cpu 1.733 total            
```

</details>

<details>
<summary>Now (much worse) `git-annex-add` on the same thing:</summary>

```
❱ time ls [0-9]* | xargs git annex add
ls --color=auto [0-9]*  0.13s user 0.12s system 0% cpu 1:12.03 total             
xargs git annex add  219.54s user 23.84s system 163% cpu 2:29.17 total           
```
vs.
```
❱ time git annex add .
git annex add .  106.18s user 14.68s system 133% cpu 1:30.45 total
```

</details>

Note, that this is annex repo version 5! Smudge filtering in version 7 slows things down, but AFAICT the relation between both invocations remains close to a factor 2.
I will run those again on the same datasets as the `datalad save` calls in above post to get a better clue how this scales and for a fair comparison, but I think it's clear already that this is the main performance issue ATM.

2. There is a little more that might be worth changing. `save` (or its helpers respectively) still use the "old" `Runner`. Due to the graphs above I did a line profiling of `Runner`'s `_get_output_online` and `_process_one_line`. There's about 20% of its runtime we could do sth about (that is: 20% on my laptop -- w/ slower file access this will become less significant as the absolute time spend on this wouldn't increase). From `line_profiler`'s output:

```
Line #               Hits                  Time  Per Hit    % Time         Line Contents
==============================================================

_get_output_online:

687         50001  152411456.0     3048.2            79.5            stdout_ = self._process_one_line(*stdout_args)
688         50001    38297474.0       765.9            20.0            stdout += stdout_

_process_one_line:

736         50002          83106.0          1.7               0.0          if line is None:        
737         50002        425591.0          8.5               0.2              lgr.log(3, "Reading line from %s", out_type)
738         50002  156584535.0    3131.6             86.1              line = {'stdout': proc.stdout, 'stderr': proc.stderr}[out_type].readline()


741         50002        121482.0          2.4               0.1          if line and log_is_callable:
742                                                                                           # Let it be processed
743         50001    23723635.0      474.5             13.0              line = log_(assure_unicode(line))
```

So, I think output processing could and should be faster. Given the overall time spending 20% of the total time (see graphs in post above) for appending to a collective string (moving shit around in memory over and over) and logging it is nothing we should be content with ;-)
However, for `save` it's probably more important to switch to `WitlessRunner` in the first place and only then have a new assessment of this aspect.

**Comment by mih** [comment]
> However, for `save` it's probably more important to switch to `WitlessRunner` in the first place and only then have a new assessment of this aspect.

Which translates to "stop using `_git_custom_command()`", which I have started in https://github.com/datalad/datalad/pull/4689, but have not completed yet.

**Comment by bpoldrack** [comment]
> Which translates to "stop using _git_custom_command()", which I have started in #4689, but have not completed yet.

Not for `save`. That's `AnnexRepo.save_add` -> `AnnexRepo._run_annex_command_json` -> `AnnexRepo._run_annex_command`. We need to pass `runner='gitwitless` down there.

**Comment by bpoldrack** [comment]
Re `git annex add FILES` vs `git annex add .`:
There's another aspect: We do the command splitting to several calls due to too long an argument list. However, `git-annex-add` allows for `--batch`. Since it's not clear to me how easy it would be to introduce some special case handling, where we actually call with a directory, it's probably worth measuring whether this would give some benefit already.

**Comment by mih** [comment]
Re git annex add FILES vs git annex add .:

We have to keep in mind that passing input arguments as such is what we came from a long time ago. And then we added kludge after kludge to deal with all the kinks (is it a submodule, is it a symlink, ...).

**Comment by bpoldrack** [comment]
Yes, agree. That's why the conclusion from all of that isn't clear to me yet.

FWIW: I'm playing with batching right now. May be this yields some easier to digest compromise.

**Comment by bpoldrack** [comment]
Update Re batched `git-annex-add`:

1. It's worth it!

    ```
    ❱ time ls *.txt | git annex add --batch
    
    [snip]
    
    ls --color=auto *.txt  0.29s user 0.17s system 0% cpu 3:59.21 total
     git annex add --batch  337.92s user 69.59s system 140% cpu 4:49.06 total
    ```
    Compare to above calls!

2. Something is broken in `BatchedAnnex(es)`:
    Stalls after 2097 files. I suspect that's due to the fact, that annex tells `(recording state in git ...)` every few thousand files.
    May be we can't handle empty line returns from batched processes correctly (which we need for other reasons, too).

Ping @mih

**Comment by mih** [comment]
FTR: https://github.com/datalad/datalad/pull/4850 closes the `_git_custom_command()` chapter.

**Comment by bpoldrack** [comment]
FTR: I'm back on it.

Implemented use of batched git-annex-add and am testing things.
This issue:

> Stalls after 2097 files.

seems to have disappeared.
Currently double-checking measurements, since quite some lo-level things have changed since.

**Comment by yarikoptic** [comment]
@bpoldrack :

> FTR: I'm back on it.
> 
> Implemented use of batched git-annex-add and am testing things.

looking at current AnnexRepo.add -- we still do not have `batch=None` option, so I guess PR is yet to be worked out,  right?

**Comment by yarikoptic** [comment]
I am not entirely sure we should keep this issue open since it is not actionable on its own and "perpetual" (since "extremely" is not quantified). I do not think we would be able to achieve some parity with plain git/git-annex invocations since we "do more". And that was my answer to a user who also observed that `save` is slower than plain `git`: https://neurostars.org/t/processing-time-of-datalad-is-slower-than-time-of-git/23372/2 . 

Some issues related to `save` were addressed on git-annex level since then and #5721 is suggested to address a particular use case of saving lots of files.

But overall -- should we establish some *goal* to aim for? e.g. "no more than 100% overhead in comparison to plain `git` invocation"? or smth like that + add a benchmark for `save_lots_of_files` (apparently we have none like that really) and `save_lots_of_subdatasets` to ensure that we are staying "on track" and do not introduce regressions. 

WDYT?

**Comment by mih** [comment]
I think we can close this. Another wave of performance updates will certainly come with updated estimates. At that point, this issues can serve as a "FOI" style anecdote.
